# Heart Attack Analysis and Prediction

## About Dataset

This dataset contains medical data of different patients having various health inicators using which we can analyze and predict the risk of heart attacks more accurately.

## Source

This dataset is present in Kaggle in the following link:
> https://www.kaggle.com/datasets/rashikrahmanpritom/heart-attack-analysis-prediction-dataset

## Data Dictionary

The dataset includes the following features:

- **Age**: Age of the patient.
- **Sex**: Sex of the patient
- **cp**: Chest pain type
  - Value 1: Typical angina
  - Value 2: Atypical angina
  - Value 3: Non-anginal pain
  - Value 4: Asymptomatic
- **trtbps**: Resting blood pressure (in mm Hg)
- **chol**: Cholesterol in mg/dl fetched via BMI sensor
- **fbs**: Fasting blood sugar > 120 mg/dl (1 = true; 0 = false)
- **rest_ecg**: Resting electrocardiographic results
  - Value 0: Normal
  - Value 1: Having ST-T wave abnormality (T wave inversions and/or ST elevation or depression of > 0.05 mV)
  - Value 2: Showing probable or definite left ventricular hypertrophy by Estes' criteria
- **thalachh**: Maximum heart rate achieved
- **exng**: Exercise-induced angina (1 = yes; 0 = no)
- **oldpeak**: Numeric Data. This represents ST depression induced by exercise relative to rest for the patients.
- **slp**: This represents the slope of the peak exercise ST segment for the patients. Values are 0,1 and 2.
- **caa**: Number of major vessels (0-3)
- **thal**: Categorical Data. The thalassemia level in blood of patients. Values are 0, 1, 2 and 3.
- **output**: Heart attack risk indicator (0 = less chance of heart attack, 1 = more chance of heart attack)

## Problem Statements

1. **Model Training**: The objective of model training is to train the model with this data.
2. **Model Evaluation**: Evaluate the performance of the model with different metrics such as accuracy, precision, recall and f1 score.
3. **Model Optimization**: The objective of model optimization is to find the optimal model with hyperparameter tuning to avoid overfitting or underfitting and improve the performace of the model.

### Load Libraries

In [1]:
# General Libraries
import pandas as pd
import numpy as np
import os
import warnings

# Preprocessing
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split

# Model
from sklearn.neighbors import KNeighborsClassifier

# Evaluation
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Hyperparameter Tuning
from sklearn.model_selection import GridSearchCV

# Model Save
import pickle

### Settings

In [18]:
# Warning
warnings.filterwarnings("ignore")

# Path
data_path = "../data"
model_path = "../models"
# csv_path = os.path.join(data_path, "heart_wo.csv")
# csv_path = os.path.join(data_path, "heart.csv")
csv_path = os.path.join(data_path, "heart_selected.csv")

### Load Data

In [19]:
df = pd.read_csv(csv_path)

In [20]:
df.head()

age  sex  cp  trtbps  chol  thalachh  exng  oldpeak  slp  caa  thall  \
0   63    1   3     145   233       150     0      2.3    0    0      1   
1   37    1   2     130   250       187     0      3.5    0    0      2   
2   41    0   1     130   204       172     0      1.4    2    0      2   
3   56    1   1     120   236       178     0      0.8    2    0      2   
4   57    0   0     120   354       163     1      0.6    2    0      2   

   output  
0       1  
1       1  
2       1  
3       1  
4       1

### Data Preparation for Training

In [21]:
# Separate the input and output features
X = df.iloc[:, : -1]
y = df.iloc[:, -1]

In [22]:
# Split Training and testing data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size= 0.2, random_state= 42)

In [23]:
# Scale the data
scaler = StandardScaler()
X_s = scaler.fit_transform(X)
X_train_s = scaler.fit_transform(X_train)
X_test_s = scaler.transform(X_test)

In [48]:
# Train a model and evaluate
def train_evaluate(model):
    # Train the model
    model.fit(X_train_s, y_train)

    # Predict Train
    y_train_pred = model.predict(X_train_s)
    # Predict test
    y_test_pred = model.predict(X_test_s)

    # Print evaluation metrics for train
    print("=" * 60)
    print("Tarining Scores")
    print("=" * 60)
    print(f"Accuracy: {accuracy_score(y_train, y_train_pred): .3f}")
    print(f"Precision: {precision_score(y_train, y_train_pred): .3f}")
    print(f"Recall: {recall_score(y_train, y_train_pred): .3f}")
    print(f"F1: {f1_score(y_train, y_train_pred): .3f}")
    # Print evaluation metrics for test
    print("=" * 60)
    print("Testing Scores")
    print("=" * 60)
    print(f"Accuracy: {accuracy_score(y_test, y_test_pred): .3f}")
    print(f"Precision: {precision_score(y_test, y_test_pred): .3f}")
    print(f"Recall: {recall_score(y_test, y_test_pred): .3f}")
    print(f"F1: {f1_score(y_test, y_test_pred): .3f}")

In [49]:
# Try KNN
knnc = KNeighborsClassifier()
train_evaluate(knnc)

Tarining Scores
Accuracy:  0.868
Precision:  0.880
Recall:  0.880
F1:  0.880
Testing Scores
Accuracy:  0.869
Precision:  0.900
Recall:  0.844
F1:  0.871


### Insights

The evaluation metrics for the KNN classifier on the Chances of Heart Attack classification task show good overall performance, especially in terms of generalization to the testing data. Here’s a detailed analysis:

#### Training Metrics (High scores):

- **Accuracy (0.87)**: The model correctly classifies **86.8%** of the training data. This indicates that the model has learned the patterns in the training data well, but there is room for improvement, especially in recall.
- **Precision (0.88)**: The model is highly precise on the training data, correctly identifying **88%** of the patients predicted to have chances of heart attack as actual positives. This means the model is making very few false positive predictions.
- **Recall (0.88)**: The recall of **88%** indicates that the model is correctly identifying **88%** of the actual positive cases (patients with chances of heart attack). However, it is missing about **12%** of the true positive cases (false negatives), which is a concern for a health-related classification task.
- **F1 Score (0.88)**: The F1 score, which balances precision and recall, is **88%**, indicating solid performance. However, it also highlights that recall is pulling down the overall score compared to the very high precision.

#### Testing Metrics (High score):

- **Accuracy (0.87)**: The model correctly classifies **86.9%** of the testing data, a strong performance and an improvement over the training accuracy. This suggests the model generalizes well to unseen data.
- **Precision (90)**: Precision is high (**90%**) on the test data, meaning that all patients the model predicts as having chances of heart attack actually have the chances.
- **Recall (0.84)**: The recall on the test set is **84.4%**, meaning the model correctly identifies **84%** of the actual positive cases. There is a room to increase recall so that fewer cases of actual chances of heart attack are missed.
- **F1 Score (0.87)**: The F1 score on the test data is **987.1%**, reflecting a well-balanced model that handles both precision and recall well.

#### Analysis of Performance:

- **High Precision, Lower Recall:** Both in the training and testing results, precision is significantly higher than recall. This indicates that while the model is very good at making correct positive predictions (few false positives), it is still missing a proportion of actual positive cases (some false negatives). In the context of cheart attack detection, high precision is critical because false positives can lead to unnecessary medical interventions. However, recall is equally important because failing to detect actual cases of heart attack (false negatives) can have serious consequences.
- **Generalization:** The model’s testing performance is strong and even slightly better than the training performance. This suggests that the model generalizes well and is not overfitting.
- **F1 Score Balance:** The F1 score, which balances precision and recall, is quite high in both training (**0.88**) and testing (**0.87**). The high F1 score suggests that the model's trade-off between precision and recall is more favorable when generalized to unseen data.
- **Room for Recall Improvement:** Although the model performs well overall, the relatively lower recall, especially on the testing data (**0.84**), indicates that there are still some missed cases of heart attack (false negatives). Given the context of heart attack classification, improving recall would likely be a priority.


### Model Optimization

Find the optimal model using hyperparameter tuning.

In [50]:
# Tune the hyperparameter using GridSearchCV and find the best parameters
def tune_hyperparameter(model, param_dict):
    # Define Hyperparameter
    gsc = GridSearchCV(estimator = model,
                      param_grid= param_dict,
                      cv=8,
                      verbose= 1,
                      scoring= "recall")
    # Train the model
    gsc.fit(X_s, y)
    
    print(f"Best Score: {gsc.best_score_}")
    return gsc.best_params_

In [51]:
# Define hyperparameter for KNN
kn_dict = {
    "n_neighbors": range(10, 25),
    "weights": ["uniform", "distance"],
    "metric": ["mikowaski", "euclidean"]
}

In [52]:
# Hyperperameter Tuning for KNN
knnht = KNeighborsClassifier()
best_params= tune_hyperparameter(knnht, kn_dict)
print(best_params)

Fitting 8 folds for each of 60 candidates, totalling 480 fits
Best Score: 0.9098214285714286
{'metric': 'euclidean', 'n_neighbors': 19, 'weights': 'uniform'}


In [53]:
# Build with best parameters
knnmodel = KNeighborsClassifier(**best_params)
train_evaluate(knnmodel)

Tarining Scores
Accuracy:  0.839
Precision:  0.813
Recall:  0.917
F1:  0.862
Testing Scores
Accuracy:  0.902
Precision:  0.906
Recall:  0.906
F1:  0.906


### Conclusions

After hyperparameter tuning of the KNN classifier for heart attack classification, the evaluation metrics are as follows:

#### Training Scores:

- **Accuracy(0.84):** The model correctly classifies **83.9%** of the samples on the training data. This shows a solid performance on the training set, though it’s not perfect, which suggests the model is not overfitting.
- **Precision (0.81):** Out of all positive (heart attack) predictions made by the model, **81.3%** were actual heart attack cases. The precision score indicates that the model produces a fair number of false positives (misclassifying non-heart attack cases as heart attacks).
- **Recall (0.92):** The model identifies **91.7%** of the actual heart attack cases. This high recall shows the model does well in minimizing false negatives (i.e., not missing heart attack cases), which is critical in medical applications.
- **F1 Score (0.86):** The F1 score of **0.862** represents a balance between precision and recall, suggesting the model maintains a strong overall performance on the training data.

#### Testing Scores:

- **Accuracy (0.90):** The model achieves **90.2%** accuracy on the test data, which is a good improvement over the training accuracy (**83.9%**). This suggests the model generalizes well and performs even better on unseen data.
- **Precision (0.906):** The precision on the test data is higher than on the training data, indicating that **90.6%** of the positive predictions are true heart attack cases. The model now produces fewer false positives compared to the training set, meaning it is more confident and correct in predicting heart attacks.
- **Recall (0.91):** The recall on the test data matches the precision, meaning the model identifies **90.6%** of the actual heart attack cases. This shows consistent performance in terms of recall across both the training and testing sets.
- **F1 Score: 0.91):** The F1 score remains high (**0.906**), indicating a great balance between precision and recall. The model performs well in both avoiding false positives and identifying true heart attack cases on the test data.

#### Summary

- **Improved Generalization:** After hyperparameter tuning, the model shows significant improvement in generalization. The test accuracy (**90.2%**) is higher than the training accuracy (**83.9%**), which is unusual but possible when the model is able to generalize well to unseen data. This could indicate that the KNN model was well-tuned to avoid overfitting.

- **Balanced Precision and Recall:** Both precision and recall are high and balanced on the test set (**0.906**). The model is now more reliable at both catching heart attack cases and minimizing false positives, making it suitable for medical applications where both metrics are important.

- **Better Performance on Testing Data:** The higher accuracy, precision, recall, and F1 score on the test set show that the model now performs better on unseen data. This could be because hyperparameter tuning found a good balance that allows the model to perform well across different distributions of the data.